File: backend/src/main/java/com/leadgrowth/service/AuthService.java
package com.leadgrowth.service;

import com.leadgrowth.dto.*;
import com.leadgrowth.entity.*;
import com.leadgrowth.repository.*;
import com.leadgrowth.security.JwtService;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final WorkspaceRepository workspaceRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;
    private final InvitationRepository invitationRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final PasswordResetTokenRepository passwordResetTokenRepository;
    private final EmailVerificationTokenRepository emailVerificationTokenRepository;
    private final UserSessionRepository userSessionRepository;

    public AuthService(
            UserRepository userRepository,
            RoleRepository roleRepository,
            WorkspaceRepository workspaceRepository,
            PasswordEncoder passwordEncoder,
            JwtService jwtService,
            AuthenticationManager authenticationManager,
            InvitationRepository invitationRepository,
            RefreshTokenRepository refreshTokenRepository,
            PasswordResetTokenRepository passwordResetTokenRepository,
            EmailVerificationTokenRepository emailVerificationTokenRepository,
            UserSessionRepository userSessionRepository
    ) {
        this.userRepository = userRepository;
        this.roleRepository = roleRepository;
        this.workspaceRepository = workspaceRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
        this.authenticationManager = authenticationManager;
        this.invitationRepository = invitationRepository;
        this.refreshTokenRepository = refreshTokenRepository;
        this.passwordResetTokenRepository = passwordResetTokenRepository;
        this.emailVerificationTokenRepository = emailVerificationTokenRepository;
        this.userSessionRepository = userSessionRepository;
    }

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("Email is already registered");
        }

        if (!request.getPassword().equals(request.getConfirmPassword())) {
            throw new IllegalArgumentException("Passwords do not match");
        }

        // Initialize User
        User user = User.builder()
                .fullName(request.getFullName())
                .email(request.getEmail().toLowerCase())
                .phone(request.getPhone())
                .password(passwordEncoder.encode(request.getPassword()))
                .status("ACTIVE")
                .isEmailVerified(false)
                .roles(new HashSet<>())
                .build();

        // 3-Step signup onboarding selection
        if ("CREATE".equalsIgnoreCase(request.getWorkspaceAction())) {
            String wsName = request.getWorkspaceName();
            if (wsName == null || wsName.trim().isEmpty()) {
                wsName = request.getFullName() + "'s Workspace";
            }
            Workspace ws = Workspace.builder()
                    .name(wsName)
                    .companyName(request.getCompanyName())
                    .inviteCode("WS-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase())
                    .slug(wsName.toLowerCase().replaceAll("[^a-z0-9]", "-") + "-" + UUID.randomUUID().toString().substring(0, 4))
                    .build();
            workspaceRepository.save(ws);
            user.setWorkspace(ws);

            Role adminRole = roleRepository.findByName("ROLE_ADMIN")
                    .orElseThrow(() -> new IllegalStateException("ROLE_ADMIN not found"));
            user.getRoles().add(adminRole);

        } else if ("JOIN".equalsIgnoreCase(request.getWorkspaceAction())) {
            Workspace ws = workspaceRepository.findByInviteCode(request.getInviteCode())
                    .orElseThrow(() -> new IllegalArgumentException("Invalid workspace invite code"));
            user.setWorkspace(ws);

            Role userRole = roleRepository.findByName("ROLE_USER")
                    .orElseThrow(() -> new IllegalStateException("ROLE_USER not found"));
            user.getRoles().add(userRole);
        } else {
            throw new IllegalArgumentException("Workspace action must be CREATE or JOIN");
        }

        userRepository.save(user);

        // Session Setup
        UserSession session = new UserSession(null, user, "127.0.0.1", "Browser - Signup Flow", LocalDateTime.now(), LocalDateTime.now(), false);
        userSessionRepository.save(session);

        // JWT Generation
        var userDetails = new org.springframework.security.core.userdetails.User(
                user.getEmail(),
                user.getPassword(),
                user.getRoles().stream()
                        .map(role -> new org.springframework.security.core.authority.SimpleGrantedAuthority(role.getName()))
                        .collect(Collectors.toList())
        );
        String jwtToken = jwtService.generateToken(userDetails);
        String refToken = createRefreshToken(user, false);

        Set<String> roles = user.getRoles().stream().map(Role::getName).collect(Collectors.toSet());
        Workspace ws = user.getWorkspace();

        return AuthResponse.builder()
                .token(jwtToken)
                .refreshToken(refToken)
                .userId(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .roles(roles)
                .workspaceId(ws != null ? ws.getId() : null)
                .workspaceName(ws != null ? ws.getName() : null)
                .workspaceSlug(ws != null ? ws.getSlug() : null)
                .inviteCode(ws != null ? ws.getInviteCode() : null)
                .availabilityStatus(user.getAvailabilityStatus())
                .build();
    }

    @Transactional
    public AuthResponse login(LoginRequest request, String ipAddress, String userAgent) {
        User user = userRepository.findByEmail(request.getEmail().toLowerCase())
                .orElseThrow(() -> new UsernameNotFoundException("Invalid credentials"));

        if ("SUSPENDED".equalsIgnoreCase(user.getStatus())) {
            throw new IllegalArgumentException("Your account has been suspended. Please contact your workspace administrator.");
        }

        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail().toLowerCase(), request.getPassword())
        );

        // Update Last Active
        user.setLastActiveAt(LocalDateTime.now());
        userRepository.save(user);

        // Log Session
        UserSession session = new UserSession(null, user, ipAddress, userAgent, LocalDateTime.now(), LocalDateTime.now(), false);
        userSessionRepository.save(session);

        var userDetails = new org.springframework.security.core.userdetails.User(
                user.getEmail(),
                user.getPassword(),
                user.getRoles().stream()
                        .map(role -> new org.springframework.security.core.authority.SimpleGrantedAuthority(role.getName()))
                        .collect(Collectors.toList())
        );
        String jwtToken = jwtService.generateToken(userDetails);
        String refToken = createRefreshToken(user, request.isRememberMe());

        Workspace ws = user.getWorkspace();
        Set<String> roles = user.getRoles().stream().map(Role::getName).collect(Collectors.toSet());

        return AuthResponse.builder()
                .token(jwtToken)
                .refreshToken(refToken)
                .userId(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .roles(roles)
                .workspaceId(ws != null ? ws.getId() : null)
                .workspaceName(ws != null ? ws.getName() : null)
                .workspaceSlug(ws != null ? ws.getSlug() : null)
                .inviteCode(ws != null ? ws.getInviteCode() : null)
                .availabilityStatus(user.getAvailabilityStatus())
                .build();
    }

    public AuthResponse getCurrentUser(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));

        // Update Last Active
        user.setLastActiveAt(LocalDateTime.now());
        userRepository.save(user);

        Workspace ws = user.getWorkspace();
        Set<String> roles = user.getRoles().stream().map(Role::getName).collect(Collectors.toSet());

        return AuthResponse.builder()
                .userId(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .roles(roles)
                .workspaceId(ws != null ? ws.getId() : null)
                .workspaceName(ws != null ? ws.getName() : null)
                .workspaceSlug(ws != null ? ws.getSlug() : null)
                .inviteCode(ws != null ? ws.getInviteCode() : null)
                .availabilityStatus(user.getAvailabilityStatus())
                .build();
    }

    @Transactional
    public AuthResponse registerInvited(RegisterInvitedRequest request) {
        Invitation invitation = invitationRepository.findByToken(request.getToken())
                .orElseThrow(() -> new IllegalArgumentException("Invalid invitation token"));

        if ("ACCEPTED".equalsIgnoreCase(invitation.getStatus())) {
            throw new IllegalArgumentException("Invitation has already been accepted");
        }

        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("Email is already registered");
        }

        if (!request.getPassword().equals(request.getConfirmPassword())) {
            throw new IllegalArgumentException("Passwords do not match");
        }

        Role role = roleRepository.findByName("ROLE_" + invitation.getRole().toUpperCase())
                .orElseThrow(() -> new IllegalStateException("Pre-assigned role " + invitation.getRole() + " not found"));

        User user = User.builder()
                .fullName(request.getFullName())
                .email(request.getEmail().toLowerCase())
                .phone(request.getPhone())
                .password(passwordEncoder.encode(request.getPassword()))
                .workspace(invitation.getWorkspace())
                .roles(new HashSet<>(List.of(role)))
                .status("ACTIVE")
                .isEmailVerified(true) // Verified through email token acceptance
                .lastActiveAt(LocalDateTime.now())
                .build();

        userRepository.save(user);

        invitation.setStatus("ACCEPTED");
        invitationRepository.save(invitation);

        // Session Setup
        UserSession session = new UserSession(null, user, "127.0.0.1", "Browser - Joined via Invite", LocalDateTime.now(), LocalDateTime.now(), false);
        userSessionRepository.save(session);

        var userDetails = new org.springframework.security.core.userdetails.User(
                user.getEmail(),
                user.getPassword(),
                user.getRoles().stream()
                        .map(r -> new org.springframework.security.core.authority.SimpleGrantedAuthority(r.getName()))
                        .collect(Collectors.toList())
        );
        String jwtToken = jwtService.generateToken(userDetails);
        String refToken = createRefreshToken(user, false);

        Set<String> roles = user.getRoles().stream().map(Role::getName).collect(Collectors.toSet());

        return AuthResponse.builder()
                .token(jwtToken)
                .refreshToken(refToken)
                .userId(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .roles(roles)
                .workspaceId(invitation.getWorkspace().getId())
                .workspaceName(invitation.getWorkspace().getName())
                .workspaceSlug(invitation.getWorkspace().getSlug())
                .inviteCode(invitation.getWorkspace().getInviteCode())
                .availabilityStatus(user.getAvailabilityStatus())
                .build();
    }

    // Refresh Token Management
    @Transactional
    public String createRefreshToken(User user, boolean rememberMe) {
        refreshTokenRepository.deleteByUser(user); // clear old token
        refreshTokenRepository.flush();
        long expirationSec = rememberMe ? 2592000 : 86400; // 30 days vs 1 day
        String token = UUID.randomUUID().toString();
        RefreshToken refreshToken = new RefreshToken(
                null,
                token,
                user,
                Instant.now().plus(expirationSec, ChronoUnit.SECONDS)
        );
        refreshTokenRepository.save(refreshToken);
        return token;
    }

    @Transactional
    public AuthResponse refreshAccessToken(String token) {
        RefreshToken refreshToken = refreshTokenRepository.findByToken(token)
                .orElseThrow(() -> new IllegalArgumentException("Refresh token not found"));

        if (refreshToken.getExpiryDate().isBefore(Instant.now())) {
            refreshTokenRepository.delete(refreshToken);
            throw new IllegalArgumentException("Refresh token has expired. Please sign in again.");
        }

        User user = refreshToken.getUser();
        if ("SUSPENDED".equalsIgnoreCase(user.getStatus())) {
            throw new IllegalArgumentException("Account is suspended.");
        }

        // Rotate token
        String newToken = UUID.randomUUID().toString();
        refreshToken.setToken(newToken);
        refreshToken.setExpiryDate(Instant.now().plus(24, ChronoUnit.HOURS));
        refreshTokenRepository.save(refreshToken);

        var userDetails = new org.springframework.security.core.userdetails.User(
                user.getEmail(),
                user.getPassword(),
                user.getRoles().stream()
                        .map(role -> new org.springframework.security.core.authority.SimpleGrantedAuthority(role.getName()))
                        .collect(Collectors.toList())
        );
        String jwtToken = jwtService.generateToken(userDetails);
        Workspace ws = user.getWorkspace();
        Set<String> roles = user.getRoles().stream().map(Role::getName).collect(Collectors.toSet());

        return AuthResponse.builder()
                .token(jwtToken)
                .refreshToken(newToken)
                .userId(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .roles(roles)
                .workspaceId(ws != null ? ws.getId() : null)
                .workspaceName(ws != null ? ws.getName() : null)
                .workspaceSlug(ws != null ? ws.getSlug() : null)
                .inviteCode(ws != null ? ws.getInviteCode() : null)
                .availabilityStatus(user.getAvailabilityStatus())
                .build();
    }

    // Password Reset
    @Transactional
    public void requestPasswordReset(String email) {
        User user = userRepository.findByEmail(email.toLowerCase())
                .orElseThrow(() -> new UsernameNotFoundException("Email address not found."));

        String token = UUID.randomUUID().toString().substring(0, 8).toUpperCase();
        PasswordResetToken resetToken = new PasswordResetToken(
                null,
                token,
                user,
                LocalDateTime.now().plusHours(2)
        );
        passwordResetTokenRepository.save(resetToken);
        System.out.println("[PASSWORD RESET] Token generated for " + email + ": " + token);
    }

    @Transactional
    public void confirmPasswordReset(PasswordResetConfirmRequest request) {
        PasswordResetToken resetToken = passwordResetTokenRepository.findByToken(request.getToken())
                .orElseThrow(() -> new IllegalArgumentException("Invalid or expired reset token"));

        if (resetToken.getExpiryDate().isBefore(LocalDateTime.now())) {
            passwordResetTokenRepository.delete(resetToken);
            throw new IllegalArgumentException("Reset token has expired");
        }

        if (!request.getNewPassword().equals(request.getConfirmPassword())) {
            throw new IllegalArgumentException("Passwords do not match");
        }

        User user = resetToken.getUser();
        user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(user);

        passwordResetTokenRepository.delete(resetToken);
    }

    // Email Verification
    @Transactional
    public void requestEmailVerification(String email) {
        User user = userRepository.findByEmail(email.toLowerCase())
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));

        String token = UUID.randomUUID().toString();
        EmailVerificationToken verificationToken = new EmailVerificationToken(
                null,
                token,
                user,
                LocalDateTime.now().plusDays(1)
        );
        emailVerificationTokenRepository.save(verificationToken);
        System.out.println("[EMAIL VERIFICATION] Token generated for " + email + ": " + token);
    }

    @Transactional
    public void verifyEmail(String token) {
        EmailVerificationToken verificationToken = emailVerificationTokenRepository.findByToken(token)
                .orElseThrow(() -> new IllegalArgumentException("Invalid verification token"));

        if (verificationToken.getExpiryDate().isBefore(LocalDateTime.now())) {
            emailVerificationTokenRepository.delete(verificationToken);
            throw new IllegalArgumentException("Verification token has expired");
        }

        User user = verificationToken.getUser();
        user.setEmailVerified(true);
        userRepository.save(user);

        emailVerificationTokenRepository.delete(verificationToken);
    }

    // Session Monitoring
    public List<UserSessionDto> getSessions(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
        return userSessionRepository.findByUser(user).stream()
                .map(s -> new UserSessionDto(s.getId(), s.getIpAddress(), s.getUserAgent(), s.getLoginTime(), s.getLastActiveTime(), s.isExpired()))
                .collect(Collectors.toList());
    }

    @Transactional
    public void revokeSession(Long sessionId, String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
        UserSession session = userSessionRepository.findByIdAndUser(sessionId, user)
                .orElseThrow(() -> new IllegalArgumentException("Session not found"));
        session.setExpired(true);
        userSessionRepository.save(session);
    }
}

File: backend-dotnet/Services/AuthService.cs
using LeadGrowth.Data;
using LeadGrowth.DTOs;
using LeadGrowth.Models;
using LeadGrowth.Security;
using Microsoft.EntityFrameworkCore;

namespace LeadGrowth.Services;

public class AuthService : IAuthService
{
    private readonly LeadGrowthDbContext _context;
    private readonly IPasswordHasher _passwordHasher;
    private readonly IJwtService _jwtService;

    public AuthService(
        LeadGrowthDbContext context,
        IPasswordHasher passwordHasher,
        IJwtService jwtService)
    {
        _context = context;
        _passwordHasher = passwordHasher;
        _jwtService = jwtService;
    }

    public async Task<AuthResponse> RegisterAsync(RegisterRequest request)
    {
        var normalizedEmail = request.Email.Trim().ToLower();

        if (await _context.Users.AnyAsync(u => u.Email == normalizedEmail))
        {
            throw new ArgumentException("Email is already registered");
        }

        if (!string.IsNullOrWhiteSpace(request.ConfirmPassword) && !string.Equals(request.Password, request.ConfirmPassword))
        {
            throw new ArgumentException("Passwords do not match");
        }

        var user = new User
        {
            FullName = request.FullName,
            Email = normalizedEmail,
            Phone = request.Phone,
            Password = _passwordHasher.HashPassword(request.Password),
            Status = "ACTIVE",
            IsEmailVerified = false,
            CreatedAt = DateTime.UtcNow
        };

        if (string.Equals("CREATE", request.WorkspaceAction, StringComparison.OrdinalIgnoreCase))
        {
            var wsName = !string.IsNullOrWhiteSpace(request.WorkspaceName)
                ? request.WorkspaceName
                : $"{request.FullName}'s Workspace";

            var ws = new Workspace
            {
                Name = wsName,
                CompanyName = request.CompanyName,
                InviteCode = "WS-" + Guid.NewGuid().ToString("N")[..8].ToUpper(),
                Slug = wsName.ToLower().Replace(" ", "-") + "-" + Guid.NewGuid().ToString("N")[..4],
                SubscriptionPlan = "PROFESSIONAL",
                MaxUsers = 25,
                MaxLeads = 10000,
                MaxStorageMb = 5000,
                CreatedAt = DateTime.UtcNow
            };

            _context.Workspaces.Add(ws);
            await _context.SaveChangesAsync();

            user.WorkspaceId = ws.Id;
            user.Workspace = ws;

            var adminRole = await _context.Roles.FirstOrDefaultAsync(r => r.Name == "ROLE_ADMIN");
            if (adminRole == null)
            {
                adminRole = new Role { Name = "ROLE_ADMIN" };
                _context.Roles.Add(adminRole);
                await _context.SaveChangesAsync();
            }
            user.Roles.Add(adminRole);
        }
        else if (string.Equals("JOIN", request.WorkspaceAction, StringComparison.OrdinalIgnoreCase))
        {
            var cleanCode = request.InviteCode?.Trim();
            if (string.IsNullOrWhiteSpace(cleanCode))
            {
                throw new ArgumentException("Workspace invite code is required to join");
            }

            var ws = await _context.Workspaces.FirstOrDefaultAsync(w => 
                w.InviteCode == cleanCode || 
                w.InviteCode.ToLower() == cleanCode.ToLower());

            if (ws == null)
            {
                throw new ArgumentException("Invalid workspace invite code");
            }

            user.WorkspaceId = ws.Id;
            user.Workspace = ws;

            var userRole = await _context.Roles.FirstOrDefaultAsync(r => r.Name == "ROLE_USER");
            if (userRole == null)
            {
                userRole = new Role { Name = "ROLE_USER" };
                _context.Roles.Add(userRole);
                await _context.SaveChangesAsync();
            }
            user.Roles.Add(userRole);
        }

        _context.Users.Add(user);
        await _context.SaveChangesAsync();

        var session = new UserSession
        {
            UserId = user.Id,
            IpAddress = "127.0.0.1",
            UserAgent = "Browser - Signup Flow",
            LoginTime = DateTime.UtcNow,
            LastActiveTime = DateTime.UtcNow,
            IsExpired = false
        };
        _context.UserSessions.Add(session);
        await _context.SaveChangesAsync();

        var jwtToken = _jwtService.GenerateToken(user);
        var refToken = await CreateRefreshTokenAsync(user, false);

        return BuildAuthResponse(user, jwtToken, refToken);
    }

    public async Task<AuthResponse> LoginAsync(LoginRequest request, string ipAddress, string userAgent)
    {
        var normalizedEmail = request.Email.Trim().ToLower();
        var user = await _context.Users
            .Include(u => u.Workspace)
            .Include(u => u.Roles)
            .FirstOrDefaultAsync(u => u.Email == normalizedEmail);

        if (user == null || !_passwordHasher.VerifyPassword(request.Password, user.Password))
        {
            throw new UnauthorizedAccessException("Invalid credentials");
        }

        if (string.Equals("SUSPENDED", user.Status, StringComparison.OrdinalIgnoreCase))
        {
            throw new ArgumentException("Your account has been suspended. Please contact your workspace administrator.");
        }

        user.LastActiveAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        var session = new UserSession
        {
            UserId = user.Id,
            IpAddress = ipAddress,
            UserAgent = userAgent,
            LoginTime = DateTime.UtcNow,
            LastActiveTime = DateTime.UtcNow,
            IsExpired = false
        };
        _context.UserSessions.Add(session);
        await _context.SaveChangesAsync();

        var jwtToken = _jwtService.GenerateToken(user);
        var refToken = await CreateRefreshTokenAsync(user, request.RememberMe);

        return BuildAuthResponse(user, jwtToken, refToken);
    }

    public async Task<AuthResponse> GetCurrentUserAsync(string email)
    {
        var normalizedEmail = email.Trim().ToLower();
        var user = await _context.Users
            .Include(u => u.Workspace)
            .Include(u => u.Roles)
            .FirstOrDefaultAsync(u => u.Email == normalizedEmail);

        if (user == null)
        {
            throw new KeyNotFoundException("User not found");
        }

        user.LastActiveAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        return BuildAuthResponse(user, null, null);
    }

    public async Task<AuthResponse> RegisterInvitedAsync(RegisterInvitedRequest request)
    {
        var invitation = await _context.WorkspaceInvites
            .Include(i => i.Workspace)
            .FirstOrDefaultAsync(i => i.Token == request.Token);

        if (invitation == null)
        {
            throw new ArgumentException("Invalid invitation token");
        }

        if (string.Equals("ACCEPTED", invitation.Status, StringComparison.OrdinalIgnoreCase))
        {
            throw new ArgumentException("Invitation has already been accepted");
        }

        var normalizedEmail = request.Email.Trim().ToLower();
        if (await _context.Users.AnyAsync(u => u.Email == normalizedEmail))
        {
            throw new ArgumentException("Email is already registered");
        }

        if (!string.Equals(request.Password, request.ConfirmPassword))
        {
            throw new ArgumentException("Passwords do not match");
        }

        var roleName = "ROLE_" + invitation.Role.ToUpper();
        var role = await _context.Roles.FirstOrDefaultAsync(r => r.Name == roleName);
        if (role == null)
        {
            role = new Role { Name = roleName };
            _context.Roles.Add(role);
            await _context.SaveChangesAsync();
        }

        var user = new User
        {
            FullName = request.FullName,
            Email = normalizedEmail,
            Phone = request.Phone,
            Password = _passwordHasher.HashPassword(request.Password),
            WorkspaceId = invitation.WorkspaceId,
            Workspace = invitation.Workspace,
            Status = "ACTIVE",
            IsEmailVerified = true,
            LastActiveAt = DateTime.UtcNow,
            CreatedAt = DateTime.UtcNow
        };
        user.Roles.Add(role);

        _context.Users.Add(user);

        invitation.Status = "ACCEPTED";
        await _context.SaveChangesAsync();

        var session = new UserSession
        {
            UserId = user.Id,
            IpAddress = "127.0.0.1",
            UserAgent = "Browser - Joined via Invite",
            LoginTime = DateTime.UtcNow,
            LastActiveTime = DateTime.UtcNow,
            IsExpired = false
        };
        _context.UserSessions.Add(session);
        await _context.SaveChangesAsync();

        var jwtToken = _jwtService.GenerateToken(user);
        var refToken = await CreateRefreshTokenAsync(user, false);

        return BuildAuthResponse(user, jwtToken, refToken);
    }

    public async Task<AuthResponse> RefreshAccessTokenAsync(string token)
    {
        var refreshToken = await _context.RefreshTokens
            .Include(r => r.User)
                .ThenInclude(u => u!.Workspace)
            .Include(r => r.User)
                .ThenInclude(u => u!.Roles)
            .FirstOrDefaultAsync(r => r.Token == token);

        if (refreshToken == null || refreshToken.User == null)
        {
            throw new ArgumentException("Refresh token not found or user invalid");
        }

        if (refreshToken.ExpiryDate < DateTime.UtcNow)
        {
            _context.RefreshTokens.Remove(refreshToken);
            await _context.SaveChangesAsync();
            throw new ArgumentException("Refresh token has expired. Please sign in again.");
        }

        var user = refreshToken.User;
        if (string.Equals("SUSPENDED", user.Status, StringComparison.OrdinalIgnoreCase))
        {
            throw new ArgumentException("Account is suspended.");
        }

        var newToken = Guid.NewGuid().ToString("N");
        refreshToken.Token = newToken;
        refreshToken.ExpiryDate = DateTime.UtcNow.AddHours(24);
        await _context.SaveChangesAsync();

        var jwtToken = _jwtService.GenerateToken(user);

        return BuildAuthResponse(user, jwtToken, newToken);
    }

    public async Task RequestPasswordResetAsync(string email)
    {
        var normalizedEmail = email.Trim().ToLower();
        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == normalizedEmail);
        if (user == null)
        {
            throw new KeyNotFoundException("Email address not found.");
        }

        var token = Guid.NewGuid().ToString("N")[..8].ToUpper();
        var resetToken = new PasswordResetToken
        {
            UserId = user.Id,
            User = user,
            Token = token,
            ExpiryDate = DateTime.UtcNow.AddHours(2)
        };

        _context.PasswordResetTokens.Add(resetToken);
        await _context.SaveChangesAsync();

        Console.WriteLine($"[PASSWORD RESET] Token generated for {email}: {token}");
    }

    public async Task ConfirmPasswordResetAsync(PasswordResetConfirmRequest request)
    {
        var resetToken = await _context.PasswordResetTokens
            .Include(r => r.User)
            .FirstOrDefaultAsync(r => r.Token == request.Token);

        if (resetToken == null)
        {
            throw new ArgumentException("Invalid or expired reset token");
        }

        if (resetToken.ExpiryDate < DateTime.UtcNow)
        {
            _context.PasswordResetTokens.Remove(resetToken);
            await _context.SaveChangesAsync();
            throw new ArgumentException("Reset token has expired");
        }

        if (!string.Equals(request.NewPassword, request.ConfirmPassword))
        {
            throw new ArgumentException("Passwords do not match");
        }

        var user = resetToken.User;
        if (user == null)
        {
            throw new ArgumentException("User associated with reset token was not found");
        }

        user.Password = _passwordHasher.HashPassword(request.NewPassword);
        _context.PasswordResetTokens.Remove(resetToken);
        await _context.SaveChangesAsync();
    }

    public async Task RequestEmailVerificationAsync(string email)
    {
        var normalizedEmail = email.Trim().ToLower();
        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == normalizedEmail);
        if (user == null)
        {
            throw new KeyNotFoundException("User not found");
        }

        var token = Guid.NewGuid().ToString("N");
        var verificationToken = new EmailVerificationToken
        {
            UserId = user.Id,
            User = user,
            Token = token,
            ExpiryDate = DateTime.UtcNow.AddDays(1)
        };

        _context.EmailVerificationTokens.Add(verificationToken);
        await _context.SaveChangesAsync();

        Console.WriteLine($"[EMAIL VERIFICATION] Token generated for {email}: {token}");
    }

    public async Task VerifyEmailAsync(string token)
    {
        var verificationToken = await _context.EmailVerificationTokens
            .Include(v => v.User)
            .FirstOrDefaultAsync(v => v.Token == token);

        if (verificationToken == null)
        {
            throw new ArgumentException("Invalid verification token");
        }

        if (verificationToken.ExpiryDate < DateTime.UtcNow)
        {
            _context.EmailVerificationTokens.Remove(verificationToken);
            await _context.SaveChangesAsync();
            throw new ArgumentException("Verification token has expired");
        }

        var user = verificationToken.User;
        if (user != null)
        {
            user.IsEmailVerified = true;
        }
        _context.EmailVerificationTokens.Remove(verificationToken);
        await _context.SaveChangesAsync();
    }

    public async Task<List<UserSessionDto>> GetSessionsAsync(string email)
    {
        var normalizedEmail = email.Trim().ToLower();
        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == normalizedEmail);
        if (user == null)
        {
            throw new KeyNotFoundException("User not found");
        }

        var sessions = await _context.UserSessions
            .Where(s => s.UserId == user.Id)
            .OrderByDescending(s => s.LastActiveTime)
            .ToListAsync();

        return sessions.Select(s => new UserSessionDto
        {
            Id = s.Id,
            IpAddress = s.IpAddress,
            UserAgent = s.UserAgent,
            LoginTime = s.LoginTime,
            LastActiveTime = s.LastActiveTime,
            IsExpired = s.IsExpired
        }).ToList();
    }

    public async Task RevokeSessionAsync(long id, string email)
    {
        var normalizedEmail = email.Trim().ToLower();
        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == normalizedEmail);
        if (user == null)
        {
            throw new KeyNotFoundException("User not found");
        }

        var session = await _context.UserSessions.FirstOrDefaultAsync(s => s.Id == id && s.UserId == user.Id);
        if (session == null)
        {
            throw new ArgumentException("Session not found");
        }

        session.IsExpired = true;
        await _context.SaveChangesAsync();
    }

    private async Task<string> CreateRefreshTokenAsync(User user, bool rememberMe)
    {
        var oldTokens = await _context.RefreshTokens.Where(r => r.UserId == user.Id).ToListAsync();
        _context.RefreshTokens.RemoveRange(oldTokens);
        await _context.SaveChangesAsync();

        var expirationDays = rememberMe ? 30 : 1;
        var token = Guid.NewGuid().ToString("N");
        var refreshToken = new RefreshToken
        {
            UserId = user.Id,
            User = user,
            Token = token,
            ExpiryDate = DateTime.UtcNow.AddDays(expirationDays)
        };

        _context.RefreshTokens.Add(refreshToken);
        await _context.SaveChangesAsync();

        return token;
    }

    private static AuthResponse BuildAuthResponse(User user, string? token, string? refreshToken)
    {
        var roles = user.Roles.Select(r => r.Name).ToHashSet();
        var ws = user.Workspace;

        return new AuthResponse
        {
            Token = token ?? string.Empty,
            RefreshToken = refreshToken ?? string.Empty,
            Type = "Bearer",
            UserId = user.Id,
            Email = user.Email,
            FullName = user.FullName,
            Designation = user.Designation,
            Bio = user.Bio,
            ProfileImage = user.ProfileImage,
            Phone = user.Phone,
            Status = user.Status,
            Roles = roles,
            WorkspaceId = ws?.Id,
            WorkspaceName = ws?.Name,
            WorkspaceSlug = ws?.Slug,
            InviteCode = ws?.InviteCode,
            AvailabilityStatus = user.AvailabilityStatus
        };
    }
}
